File: python/app/helpers/llm_cost_helper.py
"""
Pure helper for calculating the USD cost of a single LLM completion.
≈ dotnet/src/Helper/LlmCostHelper.cs

Pricing (claude-sonnet-4-6):
  Input tokens:       $3.00 / 1M tokens
  Output tokens:      $15.00 / 1M tokens
  Cache-read tokens:  $0.30 / 1M tokens (billed instead of the full input rate)
"""
from __future__ import annotations

# Pricing constants for claude-sonnet-4-6 (USD per million tokens).
_INPUT_COST_PER_M: float       = 3.00   # $3.00 / 1M input tokens
_OUTPUT_COST_PER_M: float      = 15.00  # $15.00 / 1M output tokens
_CACHE_READ_COST_PER_M: float  = 0.30   # $0.30 / 1M cache-read tokens


def calculate_claude_sonnet_cost_usd(
    input_tokens: int | None,
    cached_tokens: int,
    output_tokens: int | None,
) -> float | None:
    """
    Calculates the USD cost of the given LLM completion using claude-sonnet-4-6 pricing.
    Returns None when token counts are unavailable.
    ≈ LlmCostHelper.CalculateClaudeSonnetCostUsd()
    """
    if input_tokens is None or output_tokens is None:
        return None

    non_cached_input = max(0, input_tokens - cached_tokens)
    return (
        non_cached_input / 1_000_000 * _INPUT_COST_PER_M
        + cached_tokens  / 1_000_000 * _CACHE_READ_COST_PER_M
        + output_tokens  / 1_000_000 * _OUTPUT_COST_PER_M
    )

File: dotnet/src/Helper/LlmCostHelper.cs
/// <summary>
/// Pure static helper for calculating the USD cost of a single LLM completion
/// using claude-sonnet-4-6 pricing:
/// <list type="bullet">
///   <item>Input tokens:       $3.00 / 1M tokens</item>
///   <item>Output tokens:      $15.00 / 1M tokens</item>
///   <item>Cache-read tokens:  $0.30 / 1M tokens (billed instead of the full input rate)</item>
/// </list>
/// </summary>
public static class LlmCostHelper
{
    private const double InputCostPerToken     = 3.00  / 1_000_000.0;
    private const double OutputCostPerToken    = 15.00 / 1_000_000.0;
    private const double CacheReadCostPerToken = 0.30  / 1_000_000.0;

    /// <summary>
    /// Calculates the USD cost of the given LLM completion result using claude-sonnet-4-6 pricing.
    /// Returns null when token counts are unavailable.
    /// </summary>
    public static double? CalculateClaudeSonnetCostUsd(LlmCompletionResult llm)
    {
        if (llm.InputTokens is null || llm.OutputTokens is null)
            return null;

        var nonCachedInput = Math.Max(0, llm.InputTokens.Value - llm.CachedTokens);

        return (nonCachedInput         * InputCostPerToken)
             + (llm.OutputTokens.Value * OutputCostPerToken)
             + (llm.CachedTokens       * CacheReadCostPerToken);
    }

    /// <summary>
    /// Calculates the USD cost of the given LLM completion result and rounds it to
    /// <paramref name="decimals"/> decimal places for display (e.g. 0.52, 0.07).
    /// Returns null when token counts are unavailable.
    /// </summary>
    public static double? CalculateRoundedCostUsd(LlmCompletionResult llm, int decimals = 2)
    {
        var cost = CalculateClaudeSonnetCostUsd(llm);
        return cost.HasValue ? Math.Round(cost.Value, decimals) : null;
    }
}
